xayn-ai/src/coi/system.rs

```diff
@@ -111,34 +116,19 @@ pub(crate) fn compute_coi(
         .collect()
 }
 
-/// Creates a new CoI that is shifted towards the position of `embedding`.
-fn shift_coi_point(embedding: &Embedding, coi: &Embedding, shift_factor: f32) -> Embedding {
-    (coi.deref() * (1. - shift_factor) + embedding.deref() * shift_factor).into()
-}
-
-/// Updates the CoIs based on the given embedding. If the embedding is close to the nearest centroid
-/// (within [`Configuration.threshold`]), the centroid's position gets updated,
-/// otherwise a new centroid is created.
-fn update_coi<
-    CP: CoiPoint + CoiPointKeyPhrases + CoiPointStats,
-    F: Fn(&str) -> Result<Embedding, Error>,
->(
-    mut cois: Vec<CP>,
-    relevances: &mut Relevances,
+/// Updates the positive coi closest to the embedding or creates a new one if it's too far away.
+fn update_positive_coi(
+    mut cois: Vec<PositiveCoi>,
     embedding: &Embedding,
+    config: &Configuration,
+    relevances: &mut Relevances,
+    smbert: impl Fn(&str) -> Result<Embedding, Error>,
     candidates: &[String],
     viewed: Duration,
-    smbert: F,
-    config: &Configuration,
-) -> Vec<CP> {
+) -> Vec<PositiveCoi> {
     match find_closest_coi_mut(embedding, &mut cois, config.neighbors()) {
         Some((coi, distance)) if distance < config.threshold() => {
-            coi.set_point(shift_coi_point(
-                embedding,
-                coi.point(),
-                config.shift_factor(),
-            ));
-            coi.set_id(Uuid::new_v4().into());
```

Comment:
This is no longer needed? (just to be sure)

Reply:
correct, i also double checked this. it was convenient for the merge/sync implementation (which we don't use in the near future), but it's very inconvenient for any centralized data structure mapping coi ids to some data (which we will use for many things in the new engine).